File: go/parsers/result.go
package parsers

import (
	"capnproto.org/go/capnp/v3"

	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-serializer-capnproto/go/gen"
)

type Result struct {
	gen *gen.Result
	ex  *PayloadExpander
}

func NewResultFields(g *gen.Result, payload []byte) messages.ResultFields {
	return &Result{
		gen: g,
		ex:  &PayloadExpander{payload: payload, serializer: g.PayloadSerializerID()},
	}
}

func (r *Result) RequestID() uint64 {
	return r.gen.RequestID()
}

func (r *Result) Details() map[string]any {
	return map[string]any{}
}

func (r *Result) Args() []any {
	return r.ex.Args()
}

func (r *Result) KwArgs() map[string]any {
	return r.ex.Kwargs()
}

func (r *Result) PayloadIsBinary() bool {
	return true
}

func (r *Result) Payload() []byte {
	return r.ex.Payload()
}

func (r *Result) PayloadSerializer() uint64 {
	return r.gen.PayloadSerializerID()
}

func ResultToCapnproto(m *messages.Result) ([]byte, error) {
	msg, seg, err := capnp.NewMessage(capnp.SingleSegment(nil))
	if err != nil {
		return nil, err
	}

	result, err := gen.NewRootResult(seg)
	if err != nil {
		return nil, err
	}

	result.SetRequestID(m.RequestID())

	data, err := msg.Marshal()
	if err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypeResult, data), nil
}

func CapnprotoToResult(data, payload []byte) (*messages.Result, error) {
	msg, err := capnp.Unmarshal(data)
	if err != nil {
		return nil, err
	}

	result, err := gen.ReadRootResult(msg)
	if err != nil {
		return nil, err
	}

	return messages.NewResultWithFields(NewResultFields(&result, payload)), nil
}

File: python/wamp_msgs_capnp/parsers/result.py
import os
from pathlib import Path
from typing import Any

import capnp
from wampproto.messages.result import Result, IResultFields
from wampproto.serializers.payload import serialize_payload

from wamp_msgs_capnp.parsers import helpers

root_dir = Path(__file__).resolve().parents[1]
schema_file = os.path.join(root_dir, "schemas", "result.capnp")
result_capnp = capnp.load(str(schema_file))


class ResultFields(IResultFields):
    def __init__(self, gen, payload: bytes):
        self._gen = gen
        self._ex = helpers.PayloadExpander(payload, gen.payloadSerializerID)

    @property
    def request_id(self) -> int:
        return self._gen.requestID

    @property
    def details(self) -> dict[str, Any]:
        return {}

    @property
    def args(self) -> list[Any]:
        return self._ex.args

    @property
    def kwargs(self) -> dict[str, Any]:
        return self._ex.kwargs

    @property
    def payload_is_binary(self) -> bool:
        return True

    @property
    def payload(self) -> bytes:
        return self._ex.payload

    @property
    def payload_serializer(self) -> int:
        return self._gen.payloadSerializerID


def result_to_capnproto(r: Result) -> bytes:
    result = result_capnp.Result.new_message()

    result.requestID = r.request_id

    payload_serializer = helpers.select_payload_serializer(r.details)
    result.payloadSerializerID = payload_serializer

    if r.payload_is_binary():
        payload = r.payload
    else:
        payload = serialize_payload(payload_serializer, r.args, r.kwargs)

    packed_data = result.to_bytes_packed()

    return helpers.prepend_header(Result.TYPE, packed_data, payload)


def capnproto_to_result(data: bytes) -> Result:
    message_data, payload_data = helpers.extract_message(data)
    result_obj = result_capnp.Result.from_bytes_packed(message_data)

    return Result(ResultFields(result_obj, payload_data))
